pkg/build/oci/publish.go

```diff
@@ -33,257 +30,236 @@ import (
 		return name.Digest{}, fmt.Errorf("failed to compute digest: %w", err)
 	}
 
-	digest := name.Digest{}
+	dig := ref.Context().Digest(hash.String())
+
 	toPublish := tags
 	msg := "publish image tag"
+
 	if !shouldPushTags {
-		toPublish = []string{fmt.Sprintf("%s@%s", strings.Split(tags[0], ":")[0], h.String())}
+		toPublish = []string{dig.String()}
 		msg = "publishing image without tag (digest only)"
 	}
+
+	g, ctx := errgroup.WithContext(ctx)
 	for _, tag := range toPublish {
-		logger.Printf("%s %v", msg, tag)
-		digest, err = publishTagFromImage(ctx, image, tag, h, local, logger, remoteOpts...)
-		if err != nil {
-			return name.Digest{}, err
-		}
+		tag := tag
+		g.Go(func() error {
+			logger.Printf("%s %v", msg, tag)
+			ref, err := name.ParseReference(tag)
+			if err != nil {
+				return fmt.Errorf("unable to parse reference: %w", err)
+			}
+
+			// Write any attached SBOMs/signatures.
+			wp := writePeripherals(ref, logger, remoteOpts...)
+			g.Go(func() error {
+				return wp(ctx, image)
+			})
+
+			g.Go(func() error {
+				return remote.Write(ref, image, remoteOpts...)
+			})
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return name.Digest{}, fmt.Errorf("failed to publish: %w", err)
 	}
 
-	return digest, nil
+	return dig, nil
 }
 
-// PublishImageFromLayer convenience function that creates an image from a v1.Layer, and then publishes that.
-// Just wraps BuildImageFromLayer and PublishImage.
-// Options provided are applied either to BuildImageFromlayer or PublishImage.
-func PublishImageFromLayer(ctx context.Context, layer v1.Layer, ic types.ImageConfiguration, created time.Time, arch types.Architecture, logger log.Logger, local, shouldPushTags bool, tags []string, remoteOpts ...remote.Option) (name.Digest, oci.SignedImage, error) {
-	v1Image, err := BuildImageFromLayer(layer, ic, created, arch, logger)
+func LoadImage(ctx context.Context, image oci.SignedImage, logger log.Logger, tags []string) (name.Reference, error) {
+	hash, err := image.Digest()
+	if err != nil {
+		return name.Digest{}, err
+	}
+	localSrcTagStr := fmt.Sprintf("%s/%s:%s", LocalDomain, LocalRepo, hash.Hex)
+	localSrcTag, err := name.NewTag(localSrcTagStr)
+	if err != nil {
+		return name.Digest{}, err
+	}
+	logger.Infof("saving OCI image locally: %s", localSrcTag.Name())
+	resp, err := daemon.Write(localSrcTag, image)
 	if err != nil {
-		return name.Digest{}, nil, err
+		logger.Errorf("docker daemon error: %s", strings.ReplaceAll(resp, "\n", "\\n"))
+		return name.Digest{}, fmt.Errorf("failed to save OCI image locally: %w", err)
 	}
-	dig, err := PublishImage(ctx, v1Image, local, shouldPushTags, logger, tags, remoteOpts...)
-	return dig, v1Image, err
+	logger.Debugf("docker daemon response: %s", strings.ReplaceAll(resp, "\n", "\\n"))
+	for _, tag := range tags {
+		localDstTag, err := name.NewTag(tag)
+		if err != nil {
+			return name.Digest{}, err
+		}
+		if strings.HasPrefix(localSrcTag.Name(), fmt.Sprintf("%s/", LocalDomain)) {
```

Comment:
Will not this always be true? Line 97 has `localSrcTagStr := fmt.Sprintf("%s/%s:%s", LocalDomain...`
I was expecting that `publish` actually loaded the tags to my Docker daemon. This prevents it.